### PR TITLE
project pages

### PR DESCRIPTION
**Pull request by Machine-Maker**
Not quite done here... but just to people can take a look.

### CODE REVIEW

File: src/main/frontend/src/javascripts/pageEdit.js [hunk 1]
```diff
@@ -23,14 +23,21 @@ $(function() {
             if (parentId !== null)
                 url = '/' + projectOwner + '/' + projectSlug + '/pages/' + parent.data('slug') + '/' + slugify(pageName) + '/edit';
         }
+        var csrfValue = $('input[type=hidden][name=_csrf]');
```
Comment (MiniDigger) [inline src/main/frontend/src/javascripts/pageEdit.js:26]: do we really need to handle this manually? that sound ewww
ore didnt do it, so surely theres a better way

Comment (MiniDigger) [reply src/main/frontend/src/javascripts/pageEdit.js:26]: can just remove all this after you rebased, I fixed it on master

File: src/main/frontend/src/javascripts/pageEdit.js [hunk 2]
```diff
@@ -23,14 +23,19 @@ $(function() {
             if (parentId !== null)
                 url = '/' + projectOwner + '/' + projectSlug + '/pages/' + parent.data('slug') + '/' + slugify(pageName) + '/edit';
         }
+        // var csrfValue = $('input[type=hidden][name=_csrf]');
+        // if (csrfValue.length) {
+        //     csrfValue = csrfValue.attr('value')
+        // } else csrfValue = "";
```
Comment (MiniDigger) [inline src/main/frontend/src/javascripts/pageEdit.js:29]: Cleanup

File: src/main/java/me/minidigger/hangar/controller/PagesController.java [hunk 3]
```diff
@@ -1,45 +1,155 @@
...
+        this.projectService = projectService;
+        this.pagesSerivce = pagesSerivce;
+        this.pagesFactory = pagesFactory;
+        this.projectPageDao = projectPageDao;
+        this.routeHelper = routeHelper;
+        this.markdownService = markdownService;
+    }
+
+    @PostMapping(value = "/pages/preview", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public Object showPreview(@RequestBody String raw) throws JSONException {
```
Comment (MiniDigger) [inline src/main/java/me/minidigger/hangar/controller/PagesController.java:58]: Return type response entity

File: src/main/java/me/minidigger/hangar/controller/PagesController.java [hunk 4]
```diff
@@ -1,45 +1,155 @@
...
+        JSONObject rawJson;
+        try {
+            rawJson = new JSONObject(raw);
+            rawJson.getString("raw");
```
Comment (MiniDigger) [inline src/main/java/me/minidigger/hangar/controller/PagesController.java:62]: Is this here to catch an exception? Or why is the result of getstring ignored here?

File: src/main/java/me/minidigger/hangar/controller/PagesController.java [hunk 5]
```diff
@@ -1,45 +1,155 @@
...
+        projectPageDao.get().delete(projectPage);
+        return new RedirectView(routeHelper.getRouteUrl("projects.show", author, slug));
     }
 
     @Secured("ROLE_USER")
-    @PostMapping("/{author}/{slug}/pages/{page}/edit")
-    public Object save(@PathVariable Object author, @PathVariable Object slug, @PathVariable Object page) {
-        return null; // TODO implement save request controller
+    @PostMapping(value = {"/{author}/{slug}/pages/{page}/edit", "/{author}/{slug}/pages/{page}/{subPage}/edit"}, consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    public Object save(@PathVariable String author, @PathVariable String slug, @PathVariable String page, @PathVariable(required = false) String subPage, @RequestParam(value = "parent-id", required = false) String parentId, @RequestParam("content") String pageContent, @RequestParam("name") String newPageName) {
```
Comment (MiniDigger) [inline src/main/java/me/minidigger/hangar/controller/PagesController.java:107]: Return type

File: src/main/java/me/minidigger/hangar/controller/PagesController.java [hunk 6]
```diff
@@ -1,45 +1,155 @@
...
+        }
+        // TODO User action log
     }
 
     @Secured("ROLE_USER")
-    @GetMapping("/{author}/{slug}/pages/{page}/edit")
-    public Object showEditor(@PathVariable Object author, @PathVariable Object slug, @PathVariable Object page) {
-        return null; // TODO implement showEditor request controller
+    @GetMapping({"/{author}/{slug}/pages/{page}/edit", "/{author}/{slug}/pages/{page}/{subPage}/edit"})
+    public Object showEditor(@PathVariable String author, @PathVariable String slug, @PathVariable String page, @PathVariable(required = false) String subPage) {
```
Comment (MiniDigger) [inline src/main/java/me/minidigger/hangar/controller/PagesController.java:134]: Return type

File: src/main/java/me/minidigger/hangar/model/viewhelpers/ProjectPage.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    public void setParentId(@Nullable Long parentId) {
+        this.parentId = parentId;
+    }
+
+    public boolean isHome() {
+        return name.equals("Home") && parentId == null; // TODO check against config default
+    }
+
+//    public String getFullSlug(ProjectPage parent) {
```
Comment (MiniDigger) [inline src/main/java/me/minidigger/hangar/model/viewhelpers/ProjectPage.java:97]: Cleanup

File: src/main/java/me/minidigger/hangar/service/AuthenticationService.java [hunk 8]
```diff
@@ -91,7 +91,6 @@ private void saveSession(ApiSession session) {
     public void loginAsFakeUser() {
         String username = hangarConfig.fakeUser.getUsername();
         UsersTable userEntry = userDao.get().getByName(username);
-        System.out.println(hangarConfig.channels.getColorDefault());
```
Comment (MiniDigger) [inline src/main/java/me/minidigger/hangar/service/AuthenticationService.java:94]: I was wondering where that cyan console log was coming from, lmao

File: src/main/java/me/minidigger/hangar/controller/PagesController.java (hunk 4 above)
Comment (Machine-Maker) [reply src/main/java/me/minidigger/hangar/controller/PagesController.java:62]: Yeah, getString throws a exception that will be cause if it doesn't exist.

File: src/main/java/me/minidigger/hangar/controller/PagesController.java (hunk 5 above)
Comment (Machine-Maker) [reply src/main/java/me/minidigger/hangar/controller/PagesController.java:107]: save() returns a ResponseEntity if its a new page, or a RedirectView if its editing an existing page. so this has to be left as Object.

**Pull request merged**